File: examples/python/map_client/correlation.py
"""Per-request correlation and W3C trace context.

Two identifiers travel with every request, and they do different jobs:

* ``x-correlation-id`` is an application-owned GUID. A human can read it off
  the screen and quote it in a support request. It is telemetry only and is
  never used for authentication, authorization, or routing.

* ``traceparent`` is W3C trace context. It lets Azure Monitor establish the
  parent/child relationship between the gateway's frontend request and its
  backend dependency.

Neither replaces the other, which is why both are sent.
"""

from __future__ import annotations

import os
import uuid
from dataclasses import dataclass, field

_TRACE_VERSION = "00"
# Sampled. The reference environments run at full sampling to prove
# correlation; production deployments lower it at the gateway.
_TRACE_FLAGS = "01"


def _new_trace_id() -> str:
    """32 lowercase hex characters, per W3C Trace Context."""
    return uuid.uuid4().hex


def _new_span_id() -> str:
    """16 lowercase hex characters, per W3C Trace Context."""
    return os.urandom(8).hex()


@dataclass(frozen=True)
class RequestContext:
    """Correlation identifiers for a single invocation.

    Frozen because reusing or mutating these across requests would break the
    one-request-one-identifier property that makes the telemetry joinable.
    """

    correlation_id: str = field(default_factory=lambda: str(uuid.uuid4()))
    trace_id: str = field(default_factory=_new_trace_id)
    span_id: str = field(default_factory=_new_span_id)

    @property
    def traceparent(self) -> str:
        return f"{_TRACE_VERSION}-{self.trace_id}-{self.span_id}-{_TRACE_FLAGS}"

    def headers(self) -> dict[str, str]:
        """Correlation headers for this request.

        Note what is absent: no identity headers. The gateway derives identity
        exclusively from the validated token and strips any caller-supplied
        ``x-user-id``, ``x-tenant-id``, or similar. Sending them would be
        pointless at best and misleading at worst.
        """
        return {
            "x-correlation-id": self.correlation_id,
            "traceparent": self.traceparent,
        }

File: src/vscode/src/correlation.ts
/**
 * Per-request correlation and W3C trace context.
 *
 * One GUID and one trace per invocation. The correlation ID is telemetry
 * only: it is never used for authentication, authorization, or routing, so a
 * caller gaining control of it gains nothing.
 */

import { randomUUID, randomBytes } from 'node:crypto';

const TRACE_VERSION = '00';
const TRACE_FLAGS_SAMPLED = '01';

export interface RequestContext {
  readonly correlationId: string;
  readonly traceId: string;
  readonly spanId: string;
  readonly traceparent: string;
}

export function createRequestContext(): RequestContext {
  const correlationId = randomUUID();
  // 32 lowercase hex chars, per W3C Trace Context.
  const traceId = randomBytes(16).toString('hex');
  // 16 lowercase hex chars.
  const spanId = randomBytes(8).toString('hex');

  return {
    correlationId,
    traceId,
    spanId,
    traceparent: `${TRACE_VERSION}-${traceId}-${spanId}-${TRACE_FLAGS_SAMPLED}`,
  };
}

/**
 * Correlation headers for a request.
 *
 * Note what is absent: no identity headers. The gateway derives identity
 * solely from the validated token and strips caller-supplied `x-user-id`,
 * `x-tenant-id`, and similar. Sending them would achieve nothing.
 */
export function correlationHeaders(context: RequestContext): Record<string, string> {
  return {
    'x-correlation-id': context.correlationId,
    traceparent: context.traceparent,
  };
}
